src/GitVersionCore/Helpers/LockFile.cs

```diff
@@ -0,0 +1,208 @@
+/*
+MIT License
+
+Copyright 2020 Teroneko
+
+Permission is hereby granted, free of charge, to any person obtaining a copy 
+of this software and associated documentation files (the "Software"), to deal 
+in the Software without restriction, including without limitation the rights to 
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies 
+of the Software, and to permit persons to whom the Software is furnished to do 
+so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all 
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
+IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+using System;
+using System.IO;
+using System.Threading;
+
+namespace GitVersion.Helpers
+{
+    /// <summary>
+    /// This helper class can lock files.
+    /// </summary>
+    public static class LockFile
```

Comment:
This is API, it will be turned private and wrapped by FileLocker at your suggestion - ok? Like told it is only API it can be static, outside it won't be viewable, but then the FileLocker.

Reply:
> > I ask you, which timeout is necessary. I just took 15 seconds.
> 
> We have reports that GitVersion is taking more than a minute to complete, so I'm not sure 15 seconds is enough. [#1448 (comment)](https://github.com/GitTools/GitVersion/issues/1448#issuecomment-408936340)

See comment above.

> > I am worried about the fact, that the assemblies are often cached, but the servicecollection filled and serviceprovider built is always happening once in a task, so it will always dispose correct at end of any task in GitVersionTask.MSBuild assembly.
> 
> If I understand you correctly, that's the source of the problem and why this is so hard to fix: The lock needs to exist and work across processes. Locking inside the GitVersion process won't help, since we have many instances of GitVersion running in parallel in different processes.

The point is we have to ensure, that each Task (from MSBuild) in GitVersionTask.MSbuild assembly is correctly disposing the singleton IFileLock. (in this current state of implementation - changed later). So as you can see is IServiceProvider not implementing IDisposable, but ServiceProvider in assembly Microsoft.Extensions.DependencyInjection is doing so, we have to return a derived IServiceProvider implementation that implements IDisposable and write a wrapper like I did for IFileLock. With such exposed custom IServiceProvider we can now manually dispose [Microsoft.Extensions.DependencyInjection.ServiceProvider](https://docs.microsoft.com/de-de/dotnet/api/microsoft.extensions.dependencyinjection.servicecollectioncontainerbuilderextensions.buildserviceprovider?view=dotnet-plat-ext-3.1).